File: modal-js/examples/sandbox-filesystem.ts
import { App } from "modal";

/**
 * Example demonstrating filesystem operations in a Modal sandbox.
 *
 * This example shows how to:
 * - Open files for reading and writing
 * - Read file contents as text and binary data
 * - Write data to files
 * - Seek to specific positions
 * - Close file handles
 */

const app = await App.lookup("libmodal-example", { createIfMissing: true });
const image = await app.imageFromRegistry("alpine:3.21");

// Create a sandbox
const sb = await app.createSandbox(image);
console.log("Started sandbox:", sb.sandboxId);

try {
  // Write a file
  const writeHandle = await sb.open("/tmp/example.txt", "w");
  await writeHandle.write("Hello, Modal filesystem!\n");
  await writeHandle.write("This is line 2.\n");
  await writeHandle.write("And this is line 3.\n");
  await writeHandle.close();

  // Read the entire file as text
  const readHandle = await sb.open("/tmp/example.txt", "r");
  const content = await readHandle.read({ encoding: "utf8" });
  console.log("File content:", content);
  await readHandle.close();

  // Read specific number of bytes as text
  const partialHandle = await sb.open("/tmp/example.txt", "r");
  const firstLine = await partialHandle.read({ length: 25, encoding: "utf8" });
  console.log("First line:", firstLine);
  await partialHandle.close();

  // Append to the file
  const appendHandle = await sb.open("/tmp/example.txt", "a");
  await appendHandle.write("This line was appended.\n");
  await appendHandle.close();

  // Read from a specific position
  const seekHandle = await sb.open("/tmp/example.txt", "r");
  await seekHandle.seek(10); // Skip first 10 bytes
  const fromPosition = await seekHandle.read({ length: 15, encoding: "utf8" });
  console.log("From position 10:", fromPosition);
  await seekHandle.close();

  // Binary file operations
  const binaryHandle = await sb.open("/tmp/data.bin", "w");
  const binaryData = new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
  await binaryHandle.write(binaryData);
  await binaryHandle.close();

  // Read binary data (default encoding is 'binary')
  const readBinaryHandle = await sb.open("/tmp/data.bin", "r");
  const readData = await readBinaryHandle.read({ encoding: "binary" });
  console.log("Binary data:", Array.from(readData));
  await readBinaryHandle.close();

  // Demonstrate different encoding options
  const textHandle1 = await sb.open("/tmp/example.txt", "r");
  const textHandle2 = await sb.open("/tmp/example.txt", "r");

  // Read as binary (Uint8Array)
  const binaryContent = await textHandle1.read({ encoding: "binary" });
  console.log("Binary content length:", binaryContent.length);

  // Read as text (string)
  const textContent = await textHandle2.read({ encoding: "utf8" });
  console.log("Text content length:", textContent.length);

  await textHandle1.close();
  await textHandle2.close();
} catch (error) {
  console.error("Filesystem operation failed:", error);
} finally {
  // Clean up the sandbox
  await sb.terminate();
}

File: modal-rb/examples/sandbox_filesystem.rb
require 'modal'

def main
  app = Modal::App.lookup("sandbox-filesystem-example", create_if_missing: true)
  puts "Found app: #{app.app_id}"

  image = app.image_from_registry("alpine:latest")
  puts "Using image: #{image.image_id}"

  sandbox = app.create_sandbox(image,
    timeout: 60000,
    cpu: 0.5,
    memory: 256,
    command: ["sleep", "100"]  # Keep it alive for 100 seconds
  )
  puts "Created sandbox with ID: #{sandbox.sandbox_id}"

  # Wait a moment for the sandbox to be ready
  sleep(2)

  # Create and write to a file
  puts "\n=== Writing to a file ==="
  file = sandbox.open("/tmp/hello.txt", "w")
  file.write("Hello from Modal Ruby!\n")
  file.write("This is line 2\n")
  file.flush
  file.close
  puts "✓ Created and wrote to /tmp/hello.txt"

  # Read the file back
  puts "\n=== Reading the file back ==="
  file = sandbox.open("/tmp/hello.txt", "r")
  content = file.read
  puts "File content:"
  puts content
  file.close

  # Create a directory and file structure
  puts "\n=== Creating directory structure ==="

  # Use exec to create directories (since we don't have mkdir method yet)
  result = sandbox.exec(["mkdir", "-p", "/tmp/mydir/subdir"])
  exit_code = result.wait
  puts "✓ Created directories (exit code: #{exit_code})"

  # Create a file in the subdirectory
  file = sandbox.open("/tmp/mydir/subdir/data.txt", "w")
  file.write("Data in subdirectory\n")
  file.close
  puts "✓ Created /tmp/mydir/subdir/data.txt"

  # List directory contents
  puts "\n=== Listing directory contents ==="
  result = sandbox.exec(["ls", "-la", "/tmp/mydir/subdir"])
  output = result.stdout.read
  puts "Directory listing:"
  puts output
  result.wait

  # Append to a file
  puts "\n=== Appending to file ==="
  file = sandbox.open("/tmp/hello.txt", "a")
  file.write("This is an appended line\n")
  file.close

  # Read it back
  file = sandbox.open("/tmp/hello.txt", "r")
  updated_content = file.read
  puts "Updated file content:"
  puts updated_content
  file.close

  # Work with binary data
  puts "\n=== Working with binary data ==="
  binary_data = [0x48, 0x65, 0x6c, 0x6c, 0x6f].pack('C*')  # "Hello" in binary
  file = sandbox.open("/tmp/binary.dat", "wb")
  file.write(binary_data)
  file.close

  # Read binary data back
  file = sandbox.open("/tmp/binary.dat", "rb")
  read_binary = file.read
  puts "Binary data read back: #{read_binary.unpack('C*').inspect}"
  file.close

  # Use the filesystem with a process
  puts "\n=== Using filesystem with processes ==="

  # Create a script file
  script_content = <<~SCRIPT
    #!/bin/sh
    echo "Script started"
    echo "Current directory: $(pwd)"
    echo "Files in /tmp:"
    ls -la /tmp/
    echo "Script finished"
  SCRIPT

  file = sandbox.open("/tmp/myscript.sh", "w")
  file.write(script_content)
  file.close

  # Make it executable and run it
  sandbox.exec(["chmod", "+x", "/tmp/myscript.sh"]).wait
  result = sandbox.exec(["/tmp/myscript.sh"])
  script_output = result.stdout.read
  puts "Script output:"
  puts script_output
  result.wait

  puts "\n=== Filesystem example completed ==="

  # Clean up
  sandbox.terminate
  puts "✓ Sandbox terminated"

rescue => e
  puts "Error: #{e.message}"
  puts e.backtrace if ENV['DEBUG']
ensure
  begin
    sandbox&.terminate
  rescue => e
    puts "Failed to terminate sandbox: #{e.message}"
  end
end

main
